compiler/src/dotty/tools/dotc/transform/init/Objects.scala

```diff
@@ -487,7 +488,33 @@ object Objects:
 
   inline def cache(using c: Cache.Data): Cache.Data = c
 
-  type Contextual[T] = (Context, State.Data, Env.Data, Cache.Data, Heap.MutableData, Regions.Data, Trace) ?=> T
+
+  /**
+   * Handle return statements in methods and non-local returns in functions.
+   */
+  object Returns:
+    private class ReturnData(val method: Symbol, val values: mutable.ArrayBuffer[Value])
+    opaque type Data = mutable.ArrayBuffer[ReturnData]
+
+    def empty(): Data = mutable.ArrayBuffer()
+
+    def installHandler(meth: Symbol)(using data: Data): Unit =
+      data.addOne(ReturnData(meth, mutable.ArrayBuffer()))
+
+    def popHandler(meth: Symbol)(using data: Data): Value =
+      val returnData = data.remove(data.size - 1)
+      assert(returnData.method == meth, "Symbol mismatch in return handlers, expect = " + meth + ", found = " + returnData.method)
+      returnData.values.join
+
+    def handle(meth: Symbol, value: Value)(using data: Data, trace: Trace, ctx: Context): Unit =
+      data.findLast(_.method == meth) match
+        case Some(returnData) =>
+          returnData.values.addOne(value)
+
+        case None =>
+          report.error("[Internal error] Unhandled return for method " + meth + " in " + meth.owner.show + ". Trace:\n" + Trace.show, Trace.position)
```

Comment:
Make sure internal errors are reported as warnings so that it will not stop compilation.